File: archive/cdktf-java/Pr3239/lib/src/main/java/app/constructs/S3Construct.java
package app.constructs;

import com.hashicorp.cdktf.providers.aws.s3_bucket.S3Bucket;
import com.hashicorp.cdktf.providers.aws.s3_bucket.S3BucketConfig;
import com.hashicorp.cdktf.providers.aws.s3_bucket_public_access_block.S3BucketPublicAccessBlock;
import com.hashicorp.cdktf.providers.aws.s3_bucket_public_access_block.S3BucketPublicAccessBlockConfig;
import com.hashicorp.cdktf.providers.aws.s3_bucket_versioning.S3BucketVersioningA;
import com.hashicorp.cdktf.providers.aws.s3_bucket_versioning.S3BucketVersioningAConfig;
import com.hashicorp.cdktf.providers.aws.s3_bucket_versioning.S3BucketVersioningVersioningConfiguration;
import software.constructs.Construct;

public class S3Construct extends BaseConstruct {
    private final S3Bucket bucket;

    public S3Construct(final Construct scope, final String id) {
        super(scope, id);

        // Create S3 bucket
        this.bucket = new S3Bucket(this, "deployment-bucket", S3BucketConfig.builder()
                .bucket(resourceName("lambda-deployments-" + System.currentTimeMillis()).toLowerCase())
                .tags(getTagsWithName("DeploymentBucket"))
                .build());

        // Enable versioning
        new S3BucketVersioningA(this, "bucket-versioning", S3BucketVersioningAConfig.builder()
                .bucket(bucket.getId())
                .versioningConfiguration(S3BucketVersioningVersioningConfiguration.builder()
                        .status("Enabled")
                        .build())
                .build());

        // Block public access
        new S3BucketPublicAccessBlock(this, "bucket-pab", S3BucketPublicAccessBlockConfig.builder()
                .bucket(bucket.getId())
                .blockPublicAcls(true)
                .blockPublicPolicy(true)
                .ignorePublicAcls(true)
                .restrictPublicBuckets(true)
                .build());
    }

    public String getBucketArn() {
        return bucket.getArn();
    }

    public String getBucketName() {
        return bucket.getBucket();
    }
}

File: archive/cdktf-ts/Pr7795/lib/constructs/s3-construct.ts
import { Construct } from 'constructs';
import { S3Bucket } from '@cdktf/provider-aws/lib/s3-bucket';
import { S3BucketVersioningA } from '@cdktf/provider-aws/lib/s3-bucket-versioning';
import { S3BucketServerSideEncryptionConfigurationA } from '@cdktf/provider-aws/lib/s3-bucket-server-side-encryption-configuration';
import { S3BucketLifecycleConfiguration } from '@cdktf/provider-aws/lib/s3-bucket-lifecycle-configuration';
import { S3BucketPublicAccessBlock } from '@cdktf/provider-aws/lib/s3-bucket-public-access-block';

export interface S3ConstructProps {
  environmentSuffix: string;
  bucketName: string;
  enableVersioning?: boolean;
  lifecycleRules?: any[];
  tags?: Record<string, string>;
}

export class S3Construct extends Construct {
  public readonly bucket: S3Bucket;

  constructor(scope: Construct, id: string, props: S3ConstructProps) {
    super(scope, id);

    const {
      environmentSuffix,
      bucketName,
      enableVersioning = true,
      lifecycleRules = [],
      tags = {},
    } = props;

    // Create S3 bucket
    this.bucket = new S3Bucket(this, 'bucket', {
      bucket: `${bucketName}-${environmentSuffix}`,
      forceDestroy: true,
      tags: {
        Name: `${bucketName}-${environmentSuffix}`,
        ...tags,
      },
    });

    // Enable versioning
    if (enableVersioning) {
      new S3BucketVersioningA(this, 'versioning', {
        bucket: this.bucket.id,
        versioningConfiguration: {
          status: 'Enabled',
        },
      });
    }

    // Enable encryption
    new S3BucketServerSideEncryptionConfigurationA(this, 'encryption', {
      bucket: this.bucket.id,
      rule: [
        {
          applyServerSideEncryptionByDefault: {
            sseAlgorithm: 'AES256',
          },
          bucketKeyEnabled: true,
        },
      ],
    });

    // Configure lifecycle rules
    if (lifecycleRules.length > 0) {
      new S3BucketLifecycleConfiguration(this, 'lifecycle', {
        bucket: this.bucket.id,
        rule: lifecycleRules.map((rule, index) => ({
          id: rule.id || `rule-${index}`,
          status: 'Enabled',
          ...rule,
        })),
      });
    } else {
      // Default lifecycle rule
      new S3BucketLifecycleConfiguration(this, 'lifecycle', {
        bucket: this.bucket.id,
        rule: [
          {
            id: 'transition-to-ia',
            status: 'Enabled',
            transition: [
              {
                days: 30,
                storageClass: 'STANDARD_IA',
              },
              {
                days: 90,
                storageClass: 'GLACIER',
              },
            ],
          },
          {
            id: 'expire-old-versions',
            status: 'Enabled',
            noncurrentVersionExpiration: [
              {
                noncurrentDays: 90,
              },
            ],
          },
        ],
      });
    }

    // Block public access
    new S3BucketPublicAccessBlock(this, 'public-access-block', {
      bucket: this.bucket.id,
      blockPublicAcls: true,
      blockPublicPolicy: true,
      ignorePublicAcls: true,
      restrictPublicBuckets: true,
    });
  }
}
